### PR TITLE
Remove PREALLOCATE for HASHED/SPARSE_HASHED dictionaries

### LINKED ISSUE #41985: Excessive dictionary memory preallocating
This bug is reproduced on any version ( 22.1.3.7 , 22.3.11.12 , 22.9.3)

**How to reproduce** 
*(cant reproduce well on fiddle, but easy locally or on production)*

**NOT OK (with preallocate)**
```
CREATE TABLE default.oom_table (key String) ENGINE = MergeTree() ORDER BY tuple();

INSERT INTO default.oom_table SELECT hex(sipHash128(toString(number % 100))) FROM numbers(1000000000);

CREATE VIEW default.oom_view AS SELECT xxHash32(key) hash, key FROM default.oom_table GROUP BY key;

CREATE DICTIONARY default.oom_dict (hash UInt32, key String) PRIMARY KEY hash SOURCE(CLICKHOUSE(host 'localhost' user 'default' port 9000 db 'default' table 'oom_view')) LAYOUT (HASHED(PREALLOCATE 1)) LIFETIME (600);

SELECT * FROM dictionary('default.oom_dict');
```

**Error**
```
Code: 241. DB::Exception: Received from clickhouse-server:9000. DB::Exception: Memory limit (total) exceeded: would use 48.48 GiB (attempt to allocate chunk of 51540792506 bytes), maximum: 13.50 GiB. OvercommitTracker decision: Query was selected to stop by OvercommitTracker.. (MEMORY_LIMIT_EXCEEDED)

```

**OK (without preallocate)**
```
CREATE TABLE default.oom_table (key String) ENGINE = MergeTree() ORDER BY tuple();

INSERT INTO default.oom_table SELECT hex(sipHash128(toString(number % 100))) FROM numbers(1000000000);

CREATE VIEW default.oom_view AS SELECT xxHash32(key) hash, key FROM default.oom_table GROUP BY key;

CREATE DICTIONARY default.oom_dict (hash UInt32, key String) PRIMARY KEY hash SOURCE(CLICKHOUSE(host 'localhost' user 'default' port 9000 db 'default' table 'oom_view')) LAYOUT (HASHED()) LIFETIME (600);

SELECT * FROM dictionary('default.oom_dict');
```

### PR DESCRIPTION
**Pull request by azat**
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Remove PREALLOCATE for HASHED/SPARSE_HASHED dictionaries

### Documentation entry for user-facing changes
It does not give significant benefit, but now, you hashed/sparse_hashed dictionaries can be filled in parallel (#40003), using sharded dictionaries, and this should be used instead of PREALLOCATE.

Note, that dictionaries, that had been created with PREALLOCATE will work, but simply ignore this attribute.

Fixes: #41985 (cc @alexey-milovidov)
Reverts: #23979 (cc @kitaisreal)

Note: I've marked this as `Backward Incompatible Change` since this removes functionality, however, unlikely that it was broadly used, and as written above dictionaries that had been created with `PREALLOCATE` will simply ignore this attribute.

### CODE REVIEW

File: docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-layout.md [hunk 1]
```diff
@@ -202,7 +193,7 @@ Configuration example:
 or
 
 ``` sql
-LAYOUT(SPARSE_HASHED([PREALLOCATE 0]))
+LAYOUT(SPARSE_HASHED([]))
```
Comment (kitaisreal) [inline docs/en/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-layout.md:196]: Need to remove `[]`.

File: docs/ru/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-layout.md [hunk 2]
```diff
@@ -173,7 +164,7 @@ LAYOUT(HASHED(PREALLOCATE 0))
 или
 
 ``` sql
-LAYOUT(SPARSE_HASHED([PREALLOCATE 0]))
+LAYOUT(SPARSE_HASHED([]))
```
Comment (kitaisreal) [inline docs/ru/sql-reference/dictionaries/external-dictionaries/external-dicts-dict-layout.md:167]: Need to remove `[]`.

**Review by kitaisreal — CHANGES_REQUESTED** [review]
LGMT. Need to add test that we can create dictionary with `PREALLOCATE` setting without exception for backward compatibility.

**Comment by azat** [comment]
> Need to add test that we can create dictionary with PREALLOCATE setting without exception for backward compatibility.

Ok, I've added the test (though right now this is how dictionary attributes works, unknown are simply ignored).
BTW what do you think, maybe it worth produce a warning if a dictionary with `PREALLOCATE` was created?

**Comment by kitaisreal** [comment]
@azat I agree we can produce warning if we see PREALLOCATE.

**Review by kitaisreal — APPROVED** [review]

**Comment by kitaisreal** [comment]
@azat need to fix docs check.

**Comment by azat** [comment]
> need to fix docs check.

@kitaisreal is it a real problem? As far as I remember the difference between previous version of the patch (f5930d4d04ed5c9a790a324228dde562b8a30401) and current (4366f7fb3b02a0492def7d59e4abab570dba9738) does not contains docs changes:

```
$ git di f5930d4d04ed5c9a790a324228dde562b8a30401..4366f7fb3b02a0492def7d59e4abab570dba9738 --stat
 src/Dictionaries/HashedDictionary.cpp                            |  3 +++
 tests/queries/0_stateless/01509_dictionary_preallocate.reference |  2 +-
 tests/queries/0_stateless/01509_dictionary_preallocate.sh        | 40 ++++++++++++++++++++++++++++++++++++++++
 tests/queries/0_stateless/01509_dictionary_preallocate.sql       | 32 --------------------------------
 4 files changed, 44 insertions(+), 33 deletions(-)
```

While for previous patch docs check had been passed - https://s3.amazonaws.com/clickhouse-test-reports/45388/f5930d4d04ed5c9a790a324228dde562b8a30401/docs_check.html

Plus the check itself does not provide any details, or I'm missing something.

**Pull request merged**